File: src/EsiStatics.UnitTests/NavigationTests.fs
namespace EsiStatics.UnitTests

open System
open Xunit
open FluentAssertions
open EsiStatics

module NavigationTests=

    
    [<Theory>]
    [<InlineData(KnownSystems.adirain, KnownSystems.avenod, 14)>]
    [<InlineData(KnownSystems.deepari, KnownSystems.adirain, 18)>]
    [<InlineData(KnownSystems.deepari, KnownSystems.avenod, 18)>]
    [<InlineData(KnownSystems.deepari, KnownSystems.jita, 26)>]
    [<InlineData(KnownSystems.raeghoscon, KnownSystems.adirain, 6)>]
    [<InlineData(KnownSystems.adirain, KnownSystems.raeghoscon, 13)>] // should be 6 if symmetric
    let ``findRoute euclidean``(start, finish, expected) =
        let s = start |> knownSystem
        let f = finish |> knownSystem
        
        let result = (s, f) |> Navigation.findGateRoute Navigation.euclideanSystemDistance

        Assert.Equal(expected, result.Length)


    [<Theory>]
    [<InlineData(KnownSystems.adirain, KnownSystems.avenod, 14)>]
    [<InlineData(KnownSystems.avenod, KnownSystems.adirain, 13)>]
    [<InlineData(KnownSystems.deepari, KnownSystems.adirain, 18)>]
    [<InlineData(KnownSystems.adirain, KnownSystems.deepari, 19)>]
    [<InlineData(KnownSystems.deepari, KnownSystems.avenod, 18)>]
    [<InlineData(KnownSystems.avenod, KnownSystems.deepari, 18)>]
    [<InlineData(KnownSystems.deepari, KnownSystems.jita, 27)>]
    [<InlineData(KnownSystems.jita, KnownSystems.deepari, 27)>]
    [<InlineData(KnownSystems.raeghoscon, KnownSystems.adirain, 6)>]
    [<InlineData(KnownSystems.adirain, KnownSystems.raeghoscon, 6)>]
    let ``findRoute dijkstra``(start, finish, expected) =
        let s = start |> knownSystem
        let f = finish |> knownSystem
        
        let result = (s, f) |> Navigation.findGateRoute Navigation.dijkstraSystemDistance

        Assert.Equal(expected, result.Length)


    [<Theory>]
    [<InlineData(KnownSystems.adirain, KnownSystems.avenod, 14)>] 
    [<InlineData(KnownSystems.avenod, KnownSystems.adirain, 13)>] 
    [<InlineData(KnownSystems.deepari, KnownSystems.adirain, 25)>]
    [<InlineData(KnownSystems.adirain, KnownSystems.deepari, 26)>]
    [<InlineData(KnownSystems.deepari, KnownSystems.avenod, 37)>]
    [<InlineData(KnownSystems.avenod, KnownSystems.deepari, 37)>]
    [<InlineData(KnownSystems.deepari, KnownSystems.jita, 26)>]
    [<InlineData(KnownSystems.jita, KnownSystems.deepari, 49)>]
    let ``findRoute euclidean highsec preferred``(start, finish, expected) =
        let s = start |> knownSystem
        let f = finish |> knownSystem

        let result = (s, f) |> Navigation.findGateRoute Navigation.euclideanSystemDistancePreferHighsec

        Assert.Equal(expected, result.Length)
    
    [<Theory>]
    [<InlineData(KnownSystems.adirain, KnownSystems.avenod, 15)>]
    [<InlineData(KnownSystems.avenod, KnownSystems.adirain, 15)>]
    [<InlineData(KnownSystems.raeghoscon, KnownSystems.adirain, 6)>]
    [<InlineData(KnownSystems.adirain, KnownSystems.raeghoscon, 6)>]
    [<InlineData(KnownSystems.heild, KnownSystems.avenod, 27)>]
    [<InlineData(KnownSystems.avenod, KnownSystems.heild, 27)>]
    let ``findRoute euclidean highsec avoided``(start, finish, expected) =
        let s = start |> knownSystem
        let f = finish |> knownSystem

        let result = (s, f) |> Navigation.findGateRoute Navigation.euclideanSystemDistanceAvoidHighsec

        Assert.Equal(result.Length, expected)
    

    [<Theory>]
    [<InlineData(KnownSystems.thera, KnownSystems.jita)>]
    let ``findRoute on WH returns empty``(start, finish) =
        let s = start |> knownSystem
        let f = finish |> knownSystem

        let result = (s, f) |> Navigation.findGateRoute Navigation.euclideanSystemDistance

        Assert.Equal(result.Length, 0)

    
    [<Theory>]
    [<InlineData(KnownSystems.jita, KnownSystems.jita)>]
    let ``findRoute on self returns empty``(start, finish) =
        let s = start |> knownSystem
        let f = finish |> knownSystem

        let result = (s, f) |> Navigation.findGateRoute Navigation.euclideanSystemDistance

        Assert.Equal(result.Length, 0)

File: src/EsiStatics.CSharp.UnitTests/NavigationTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Xunit;
using EsiStatics;
using FluentAssertions;

namespace EsiStatics.CSharp.UnitTests
{
    public class NavigationTests
    {
        
        [Theory]
        [InlineData(0, 0, 0)]
        [InlineData(1, 1, 1)]
        [InlineData(10, 10, 10)]
        [InlineData(0, 5, 10)]
        public void Euclidean_SamePointe(float x, float y, float z)
        {
            var p1 = Position.FromCoordinates(x, y, z);
            var p2 = Position.FromCoordinates(x, y, z);

            var distance = Navigation.GetEuclidean(p1, p2);

            distance.Should().Be(0);
        }

        [Theory]
        [InlineData(1, 1, 1)]
        [InlineData(10, 10, 10)]
        [InlineData(0, 5, 10)]
        public void Euclidean_FromZero(float x, float y, float z)
        {
            var p1 = Position.Empty;
            var p2 = Position.FromCoordinates(x, y, z);

            var distance = Navigation.GetEuclidean(p1, p2);

            distance.Should().BeGreaterThan(0); 
        }


        [Theory]
        [InlineData(30005003, 30002089, 14)]
        [InlineData(30013489, 30005003, 18)]
        [InlineData(30013489, 30002089, 20)]
        [InlineData(30013489, 30000142, 10)]
        public void findRoute_euclidean(int start, int finish, int expected)
        {
            var s = SolarSystems.ById(start).Value;
            var f = SolarSystems.ById(finish).Value;
            
            var result = Navigation.FindDirectRoute(s, f);

            Assert.Equal(result.Length, expected);
        }
    }
}
